crates/ty_python_semantic/src/types.rs

```diff
@@ -4727,7 +4728,36 @@ impl<'db> Type<'db> {
                 // TODO: Use an opt-in rule for a bare `Callable`
                 KnownInstanceType::Callable => Ok(Type::Callable(CallableType::unknown(db))),
 
-                KnownInstanceType::TypingSelf => Ok(todo_type!("Support for `typing.Self`")),
+                KnownInstanceType::TypingSelf => {
+                    let scope = context.scope_id;
+                    let Some(class_ty) = context.index.enclosing_class_symbol(db, scope) else {
+                        return Err(InvalidTypeExpressionError {
+                            fallback_type: Type::unknown(),
+                            invalid_expressions: smallvec::smallvec![
+                                InvalidTypeExpression::InvalidType(*self)
+                            ],
+                        });
+                    };
+                    let class = class_ty.expect_class_type(db);
+                    let Some(class_def) = class_ty.definition(db) else {
+                        debug_assert!(false, "enclosing_class_symbol must return a class type");
+                        return Ok(Type::unknown());
+                    };
+                    let TypeDefinition::Class(d) = class_def else {
+                        debug_assert!(false, "class type must have a definition");
+                        return Ok(Type::unknown());
+                    };
+                    Ok(Type::TypeVar(TypeVarInstance::new(
+                        db,
+                        class.name(db),
```

Comment:
As mentioned above I think we can just use `Self` here -- if we need further disambiguation (like mypy and pyright provide) we would do that for all typevars.